tasks/notification_orchestrator.py

```diff
@@ -0,0 +1,631 @@
+from helpers.github_installation import get_installation_name_for_owner_for_task
+from helpers.save_commit_error import save_commit_error
+from services.comparison import (
+    get_or_create_comparison,
+)
+from services.lock_manager import LockManager, LockRetry, LockType
+from services.processing.state import ProcessingState, should_trigger_postprocessing
+from services.processing.types import ProcessingResult
+from services.redis import Redis, get_redis_connection
+from services.report import ReportService
+from services.repository import (
+    get_repo_provider_service,
+)
+from services.yaml import read_yaml_field
+from tasks.base import BaseCodecovTask
+from tasks.upload_processor import UPLOAD_PROCESSING_LOCK_NAME
+
+log = logging.getLogger(__name__)
+regexp_ci_skip = re.compile(r"\[(ci|skip| |-){3,}\]")
+
+
+class ShouldCallNotifyResult(Enum):
+    NOTIFY = "notify"
+    NOTIFY_ERROR = "notify_error"
+    DO_NOT_NOTIFY = "do_not_notify"
+    WAIT_TO_NOTIFY = "wait_to_notify"
+
+
+class ShouldCallNotifyResponse(TypedDict):
+    notification_result: ShouldCallNotifyResult
+    reason: str
+    # This is currently not used, but my idea here was to provide it and log this message in switch statements.
+    # Logging could become a bit less flexible but there's less logging altogether. Thoughts?
+    message: str
+
+
+# TODO: Ask: does this need a throws = (SoftTimeLimitExceeded,)?
+class NotificationOrchestratorTask(
+    BaseCodecovTask, name=notification_orchestrator_task_name
+):
+    def run_impl(
+        self,
+        db_session: Session,
+        *args,
+        repoid: int,
+        commitid: str,
+        commit_yaml: dict,
+        processing_results: list[ProcessingResult],
+        report_code: str | None = None,
+        empty_upload=None,
+        **kwargs,
+    ):
+        lock_manager = LockManager(
+            repoid=repoid,
+            commitid=commitid,
+            # Currently hardcoded to coverage, should be variable to the report type
+            report_type=ReportType.COVERAGE,
+            # TODO: Ask if we need a specific timeout hard timeout for this task, unsure
+            lock_timeout=max(80, self.hard_time_limit_task),
+        )
+        try:
+            lock_acquired = False
+            with lock_manager.locked(
+                lock_type=LockType.NOTIFICATION_ORCHESTRATOR,
+                retry_num=self.request.retries,
+            ):
+                lock_acquired = True
+                return self.run_impl_within_lock(
+                    db_session,
+                    repoid=repoid,
+                    commitid=commitid,
+                    commit_yaml=commit_yaml,
+                    processing_results=processing_results,
+                    report_code=report_code,
+                    empty_upload=empty_upload,
+                    **kwargs,
+                )
+        # TODO: What should happen when there's a lock error? I think this should change to some retrying mechanism
```

Comment:
I'm thinking when 2 finishers are done (presumably with different results), one picks up this lock and the other will wait. I suspect the 2nd one should be able to pick up this task